lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -141,93 +146,174 @@
+
+    def __init__(self, module):
+
+        self.m = module
+
+    @abstractmethod
+    def test(self):
+        pass
+
+    @abstractmethod
+    def list_installed(self):
+        pass
+
+    @abstractmethod
+    def get_package_details(self):
+        pass
+
+    def get_packages(self):
+
+        installed_packages = {}
+        for package in self.list_installed():
+            package_details = self.get_package_details(package)
+            package_details['source'] = self.name
+            name = package_details['name']
+            if name not in installed_packages:
+                installed_packages[name] = [package_details]
+            else:
+                installed_packages[name].append(package_details)
+        return installed_packages
+
+
+class LibMgr(PkgMgr):
+
+    LIB = None
+
+    def __init__(self, module):
+
+        super(LibMgr, self).__init__(module)
+        self.lib = None
+
+    def test(self):
+        found = False
+        try:
+            self.lib = __import__(self.LIB)
+            found = True
+        except ImportError:
+            pass
+        return found
+
+
+class CLIMgr(PkgMgr):
+
+    CLI = None
+
+    def __init__(self, module):
+
+        super(CLIMgr, self).__init__(module)
+        self.cli = None
+
+    def test(self):
+        self.cli = self.m.get_bin_path(self.CLI, False)
+        return bool(self.cli)
+
+
+class RPM(LibMgr):
+
+    name = 'rpm'
+    LIB = 'rpm'
+
+    def list_installed(self):
+        return self.lib.TransactionSet().dbMatch()
+
+    def get_package_details(self, package):
+        return dict(name=package[self.lib.RPMTAG_NAME],
+                    version=package[self.lib.RPMTAG_VERSION],
+                    release=package[self.lib.RPMTAG_RELEASE],
+                    epoch=package[self.lib.RPMTAG_EPOCH],
+                    arch=package[self.lib.RPMTAG_ARCH],)
+
+
+class APT(LibMgr):
+
+    name = 'apt'
+    LIB = 'apt'
+
+    def test(self):
+        works = super(APT, self).test()
+        if works:
+            self._cache = self.lib.Cache()
+        return works
+
+    def list_installed(self):
+        return [pk for pk in self._cache.keys() if self._cache[pk].is_installed]
+
+    def get_package_details(self, package):
+        ac_pkg = self._cache[package].installed
+        return dict(name=package, version=ac_pkg.version, arch=ac_pkg.architecture)
+
+
+class PORTAGE(CLIMgr):
+
+    name = 'portage'
+    CLI = 'qlist'
+    atoms = ['category', 'name', 'version', 'ebuild_revision', 'slots', 'prefixes', 'sufixes']
+
+    def list_installed(self):
+        rc, out, err = self.m.run_command(' '.join([self.cli, '-Iv', '|', 'xargs', '-n', '1024', 'qatom']), use_unsafe_shell=True)
+        if rc != 0 or err:
+            raise Exception("Unable to list packages rc=%s : %s" % (rc, err))
+        return out.splitlines()
+
+    def get_package_details(self, package):
+
+        pkg = {}
+        fields = package.split()
+        for i, value in enumerate(fields):
```

Comment:
```suggestion
```

Reply:
If you still want explicit loop it'd be:
```suggestion
        for key, value in zip(self.atoms, fields):
                pkg[key] = value
```

Reply:
Or

```suggestion
        return dict((key, value) for key, value in zip(self.atoms, fields))
```